File: creational/abstract-factory/python/concrete_product.py
from product import Logger, ConfigReader


class ConsoleLogger(Logger):
    def log(self, message):
        print(f"[Console Logger] {message}")


class FileLogger(Logger):
    def log(self, message):
        with open("log.txt", "a") as file:
            file.write(f"[File Logger] {message}\n")


class ConsoleConfigReader(ConfigReader):
    def read_config(self):
        config_data = input("[Config Reader] Enter config: ")
        print(f"[Config Reader] config is: {config_data}")


class FileConfigReader(ConfigReader):
    def read_config(self):
        try:
            with open("config.txt", "r") as file:
                config_data = file.read()
                print(f"[Config Reader] {config_data}")
        except FileNotFoundError:
            print("[Config Reader] Config file not found.")

File: creational/abstract-factory/typescript/concrete-product.ts
import * as fs from 'fs';
import readline from 'readline';
import { Logger, ConfigReader } from './product';

const rl = readline.createInterface({
	input: process.stdin,
	output: process.stdout,
});

export class ConsoleLogger implements Logger {
	log(message: string): void {
		console.log(`[Console Logger] ${message}`);
	}
}

export class FileLogger implements Logger {
	log(message: string): void {
		fs.appendFileSync('log.txt', `[File Logger] ${message}\n`);
	}
}

export class ConsoleConfigReader implements ConfigReader {
	readConfig(): void {
		rl.question('[Config Reader] Enter config: \n', (configData: string) => {
			console.log(`[Config Reader] Config is: ${configData}`);
			rl.close();
		});
	}
}

export class FileConfigReader implements ConfigReader {
	readConfig(): void {
		try {
			const configData = fs.readFileSync('config.txt', 'utf-8');
			console.log(`[Config Reader] ${configData}`);
		} catch (error) {
			console.log('[Config Reader] Config file not found.');
		}
	}
}
